File: FinalAssessment_19459138/Question5/FA_Graph.py
#
# DSA Final Assessment Question 5 - FA_Graph.py
#
# Name : 
# ID   :
#
# 
class FA_ListNode():
    def __init__(self, val):
        self.val = val
        self._next = None
    
    def __str__(self):
        return str(self.val)
    
class FA_LinkedList():
    def __init__(self):
        self.head = None
        self.tail = None
        self.count = 0
        
    def insertLast(self, value):
        self.count = self.count + 1
        if self.head == None:
            self.head = FA_ListNode(value)
            self.tail = self.head
        else:    
            self.tail._next = FA_ListNode(value)
            self.tail = self.tail._next
        
    def __str__(self):
        returnStr = ""
        for l in self:
            thisone = l.label
            ##print(thisone)
            returnStr = returnStr + " " + thisone
        return returnStr
        
    def __iter__(self):
        self.cursor = self.head
        return self
        
    def __next__(self):
        curval = None
        if self.cursor == None:
            raise StopIteration()
        else: 
            curval = self.cursor.val  # dumb error was here
            self.cursor = self.cursor._next
        return curval
  
        
class FA_Vertex():
    def __init__(self, label):
        self.links = FA_LinkedList()
        self.label = label
        self.count = 0
        #self.contents = contents
        self.visited = False

    def getLabel(self):
        return self.label
    
    def getValue(self):
        return self.contents
    
    def getAdjacent(self):
        return self.links
    
    def addEdge(self, vertex):
        self.links.insertLast(vertex)
        
    def getVisited(self):
        return self.visited

    def setVisited(self):
        self.visited = True

    def clearVisited(self):
        self.visited = False
    
    def __str__(self):
        return self.label + " links: " + str(self.links)
    
        
class FA_Graph():

    def __init__(self):
        self.vertices = FA_LinkedList()
        self.count = 0

    def addVertex(self, label):
        if not self.hasVertex(label):
            self.vertices.insertLast(FA_Vertex(label))

    def addEdge(self, label1, label2):
        self.addVertex(label1)  # won't add if already there
        self.addVertex(label2)
        v1 = self.getVertex(label1)
        v2 = self.getVertex(label2)
        v1.addEdge(v2)
        #v2.addEdge(v1)
        
    def getVertex(self, label):   ## Not on slides
        returnval = None
        for v in self.vertices:
            if v.label == label:
                returnval = v
        return returnval
        
    def hasVertex(self, label):
        returnval = False
        for v in self.vertices:
            if v.label == label:
                returnval = True
        return returnval
        
    def getVertexCount(self):
        return self.vertices.count
    

    def displayAsMatrix(self): 
        ...
            
    def displayWeightMatrix(self):
        ...
        

File: FinalAssessment_19459138/Question5/FA_Graph.java
/**
 * DSA Final Assessment Question 5 - FA_Graph.java                             4
 *
 * Name : Connor Kuljis
 * ID   : 19459138
 *
 **/
import java.util.*;

public class FA_Graph 
{
    private FA_LinkedList vertices;
    private FA_LinkedList edges;
    private int vertexCount;
    private int edgeCount;

    public FA_Graph() 
    {
        vertices = new FA_LinkedList();
        edges = new FA_LinkedList();
        vertexCount = 0;
        edgeCount = 0;
    }

    public void addVertex(String label, Object value) 
    {
        FA_GraphVertex vertex = new FA_GraphVertex(label, value); 
        if (!(hasVertex(label))) 
        {
            vertices.insertLast(vertex);
            vertexCount++;
        }
    }

    public void addEdge(String label1, String label2, int weight)
    {
        FA_GraphVertex v1, v2; 
        
        v1 = getVertex(label1); 
        v2 = getVertex(label2);   

	String label = "(" + label1 + "," + label2 + ")";
	FA_GraphEdge edge = new FA_GraphEdge(v1, v2, label, weight);

	edges.insertLast(edge);
	v1.addEdge(edge); // try this maybe

        edgeCount++;
    }

    public boolean hasVertex(String label) 
    {
        boolean has = false;
        for (Object v : vertices) 
        {
		   FA_GraphVertex vg = (FA_GraphVertex) v;
           if (vg.getLabel().equals(label))
			   has = true;
        }
        return has;
    }

    public FA_GraphVertex getVertex(String label) 
    {
        FA_GraphVertex theVertex = null;
        for (Object v : vertices) 
        {
		   FA_GraphVertex vg = (FA_GraphVertex) v;
           if (vg.getLabel().equals(label))
			   theVertex = vg;
        }
		return theVertex;    
    }

    public void displayAsList() 
    {
	System.out.println("\n*** DISPLAY AS LIST ***\n");
	for (Object v : vertices)
	{
	    FA_GraphVertex vertex = (FA_GraphVertex) v;

	    FA_LinkedList links = vertex.getLinks();

	    if(!links.isEmpty())
	    {
		System.out.print(vertex.getLabel() + " | " ); 
		for (Object g : links)
		{
		    FA_GraphEdge edge = (FA_GraphEdge) g;
		    System.out.print(edge + " "); 
		}
		System.out.println(""); 
	    }
	}
    }

    public void displayAsMatrix()
    {
	System.out.println("\n***DISPLAY AS MATRIX***\n"); 
	// print column headings 
	System.out.print("\t"); 
	for (Object v : vertices)
	{
	    FA_GraphVertex vertex = (FA_GraphVertex) v;
	    System.out.print(vertex.getLabel() + "\t"); 
	}
	System.out.println(""); 
	// end heading
	
	// print adjusted size ruler
	for(int i = 0; i < vertices.getSize(); i++)
	{
	    System.out.print("========"); 
	}
	System.out.println(""); 
	// end ruler

	for (Object v : vertices) // for every vertex (eg row)
	{
	    FA_GraphVertex vertex = (FA_GraphVertex) v;
	    System.out.print(vertex.getLabel() + "\t|"); // print the label
	    for(Object e : vertex.getLinks()) // for every edge in the list
	    {
		FA_GraphEdge edge = (FA_GraphEdge) e; 
		for (Object vx : vertices)  // for every vertex in the edge
		{
		    FA_GraphVertex cmpvertex = (FA_GraphVertex) vx;
		    // if the vertex in the edge matches the current vertex
		    if (((edge.getTo().getLabel()).equals(cmpvertex.getLabel()))) 
		    {
			System.out.print("1\t");  // print a match
		    }
		    else
		    {
			System.out.print("0\t");  // just print a zero
		    }
		}
	    }
	    System.out.println(""); 
	}
    }  

    public void displayWeightMatrix() // same structure as display as matrix method
    {
	System.out.println("\n***DISPLAY WEIGHT MATRIX***\n"); 
	System.out.print("\t"); 
	for (Object v : vertices)
	{
	    FA_GraphVertex vertex = (FA_GraphVertex) v;
	    System.out.print(vertex.getLabel() + "\t"); 
	}
	System.out.println(""); 
	for(int i = 0; i < vertices.getSize(); i++)
	{
	    System.out.print("========"); 
	}
	System.out.println(""); 

	for (Object v : vertices)
	{
	    FA_GraphVertex vertex = (FA_GraphVertex) v;
	    System.out.print(vertex.getLabel() + "\t|"); 
	    for(Object e : vertex.getLinks())
	    {
		FA_GraphEdge edge = (FA_GraphEdge) e;
		for (Object vx : vertices)
		{
		    FA_GraphVertex cmpvertex = (FA_GraphVertex) vx;
		    if (((edge.getTo().getLabel()).equals(cmpvertex.getLabel())))
		    {
			System.out.print(edge.getValue() + "\t");  // only difference, just print the weight here instead
		    }
		    else
		    {
			System.out.print("0\t"); 
		    }
		}
	    }
	    System.out.println(""); 
	}
    }  
}
